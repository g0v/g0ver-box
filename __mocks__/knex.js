import knex from 'jest-mock-knex';

export { client } from 'jest-mock-knex';

export default knex;

process.setMaxListeners(0);
