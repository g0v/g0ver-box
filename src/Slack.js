import _ from 'lodash';
import slack from 'slack';

export const bot = slack.rtm.client();

const TOKEN = process.env.SLACK_TOKEN;

if (TOKEN) {
  bot.listen({ token: TOKEN });
}

bot.hello((message) => {
  console.log(`Got a message: ${message.type}`);
});

const call = (name, req = {}) => (
  new Promise((resolve, reject) => {
    const fn = _.get(slack, name);
    fn(
      { ...req, token: TOKEN },
      (err, res) => (err ? reject(err) : resolve(res))
    );
  })
);

export default {
  userInfo: async req => await call('users.info', req),
  postMessage: async req => await call('chat.postMessage', { ...req, as_user: true }),
  postMultiMessage: async (users, message) => _.map(users, user => bot.ws.send(JSON.stringify({
    id: `${user}::${Date.now()}`, type: 'message', channel: user, text: message,
  }))),
  channelInfo: async req => await call('channels.info', req),
  channelJoin: async req => await call('channels.join', req),
};
