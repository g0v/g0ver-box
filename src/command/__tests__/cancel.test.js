import { client } from 'knex';
import Slack from '../../Slack';
import index from '../';

describe('cancel command', () => {
  describe('cancel <name>', () => {
    const data = { channel: 'D100', user: 'U03B2AB13', name: 'yutin', text: 'cancel 基礎建設松' };

    it('when event does not exist', async () => {
      client.mockReturnValueOnce([]);

      await index(data);
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Sorry! 找不到活動',
      });

      expect(client).toHaveBeenCalledTimes(1);
      expect(client).toMatchSnapshot();
    });

    it('permission denied', async () => {
      client.mockReturnValueOnce([{ id: '5', userId: 'U03B2AB00' }]);

      await index(data);
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Sorry! 沒有權限取消 基礎建設松 活動',
      });

      expect(client).toHaveBeenCalledTimes(1);
      expect(client).toMatchSnapshot();
    });

    it('successfully deleted', async () => {
      client.mockReturnValueOnce([{ id: '5', userId: 'U03B2AB13' }]);
      client.mockReturnValueOnce([]);

      await index(data);
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Done, 取消 基礎建設松 活動',
      });

      expect(client).toHaveBeenCalledTimes(2);
      expect(client).toMatchSnapshot();
    });
  });

  describe('remove [name]', () => {
    const data = { channel: 'D100', user: 'U03B2AB13', name: 'yutin', text: 'cancel' };

    it('when event does not exist', async () => {
      client.mockReturnValueOnce([]);

      await index(data);

      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100',
        text: 'Sorry! 找不到活動，快來揪松吧',
      });

      expect(client).toHaveBeenCalledTimes(1);
      expect(client).toMatchSnapshot();
    });

    describe('when event is existed', () => {
      beforeEach(async () => {
        client.mockReturnValueOnce([
          { id: '3', title: '基礎建設松' },
          { id: '5', title: 'event 0001' },
          { id: '6', title: 'event 0002' },
          { id: '9', title: 'event 0003' },
        ]);

        await index(data);
      });

      it('when not found event', async () => {
        await index({ ...data, text: '0' });
        expect(Slack.postMessage).toHaveBeenLastCalledWith({
          channel: 'D100',
          text: 'Sorry! 找不到活動',
        });

        expect(client).toHaveBeenCalledTimes(1);
        expect(client).toMatchSnapshot();
      });

      it('successfully deleted', async () => {
        client.mockReturnValueOnce([]);

        expect(Slack.postMessage).toHaveBeenLastCalledWith({
          channel: 'D100',
          text: expect.stringContaining('請選擇要取消的活動（輸入代碼）？'),
        });

        await index({ ...data, text: '1' });
        expect(Slack.postMessage).toHaveBeenLastCalledWith({
          channel: 'D100',
          text: 'Done, 取消 基礎建設松 活動',
        });

        expect(client).toHaveBeenCalledTimes(2);
        expect(client).toMatchSnapshot();
      });
    });
  });
});
