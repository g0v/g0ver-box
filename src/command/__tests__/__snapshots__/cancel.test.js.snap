// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cancel command cancel <name> permission denied 1`] = `
Array [
  "select * from event where deleted_at is null and title = 基礎建設松 limit 1",
]
`;

exports[`cancel command cancel <name> successfully deleted 1`] = `
Array [
  "select * from event where deleted_at is null and title = 基礎建設松 limit 1",
  "update event set deleted_at = DATE where deleted_at is null and id = 5",
]
`;

exports[`cancel command cancel <name> when event does not exist 1`] = `
Array [
  "select * from event where deleted_at is null and title = 基礎建設松 limit 1",
]
`;

exports[`cancel command remove [name] when event does not exist 1`] = `
Array [
  "select * from event where deleted_at is null and user_id = U03B2AB13 limit 1000",
]
`;

exports[`cancel command remove [name] when event is existed successfully deleted 1`] = `
Array [
  "select * from event where deleted_at is null and user_id = U03B2AB13 limit 1000",
  "update event set deleted_at = DATE where deleted_at is null and id = 3",
]
`;

exports[`cancel command remove [name] when event is existed when not found event 1`] = `
Array [
  "select * from event where deleted_at is null and user_id = U03B2AB13 limit 1000",
]
`;
